pilot/cmd/pilot-discovery/server/server_test.go

```diff
@@ -0,0 +1,145 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server_test
+
+import (
+	"io/ioutil"
+	"net"
+	"net/http"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"strings"
+
+	"github.com/golang/protobuf/ptypes"
+	"istio.io/istio/pilot/cmd/pilot-discovery/server"
+	"istio.io/istio/pilot/proxy/envoy"
+)
+
+type env struct {
+	stop   chan struct{}
+	fsRoot string
+}
+
+func (e *env) setup() (int, error) {
+	e.fsRoot = createTempDir()
+	e.stop = make(chan struct{})
+
+	port, err := getFreePort()
+	if err != nil {
+		return 0, err
+	}
+
+	// Create a test pilot discovery service configured to watch the tempDir.
+	args := server.PilotArgs{
+		Namespace: "testing",
+		DiscoveryOptions: envoy.DiscoveryServiceOptions{
+			Port:            port,
+			EnableCaching:   true,
+			EnableProfiling: true,
+		},
+		Mesh: server.MeshArgs{
+			MixerAddress:    "istio-mixer.istio-system:9091",
+			RdsRefreshDelay: ptypes.DurationProto(10 * time.Millisecond),
+		},
+		Config: server.ConfigArgs{
+			FileDir: e.fsRoot,
+		},
+		Service: server.ServiceArgs{
+			// Using the Mock service registry, which provides the hello and world services.
+			Registries: []string{string(server.MockRegistry)},
+		},
+	}
+
+	// Create and setup the controller.
+	s, err := server.NewServer(args, e.stop)
+	if err != nil {
+		return 0, err
+	}
+
+	// Start the server.
+	s.Run()
+	return port, nil
+}
+
+func (e *env) teardown() {
+	close(e.stop)
+
+	// Remove the temp dir.
+	os.RemoveAll(e.fsRoot)
+}
+
+// TODO(nmittler): getFreePort is copied from https://github.com/phayes/freeport
```

Comment:
@rshriram Is it alright to add freeport as a dependency? If so, what is the process for adding external dependencies to our build?

Reply:
You need to add to some bazel build file. And update dep’s Gopkg.lock by running dep ensure

Reply:
`getFreePort` looks racy since the listener fd is destroyed on function exit (via `l.Close()`). 

Alternatively, You can let the OS pick the port by passing `0` directly to 
https://github.com/istio/istio/pull/2003/files/31ee494b38ae1ea3dcac6b4a8a065b7b18ddc96c#diff-b5799ab52f0fa38f35765b08ddcce26bR51 and modify DiscoveryService.Run to invoke `Listen` and `Serve` separately such that you can save the port selected by `Listen` before invoking `Serve`. Then you can use the saved port in subsequent unit tests.

Reply:
@ayj good suggestion ... I've changed the discovery service's `Run` to `Start` that returns the listening address.  That obviates the need for the free port picker.